pkg/reconciler/subscription/subscription_test.go

```diff
@@ -39,6 +38,8 @@ import (
 	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
 	"github.com/knative/pkg/controller"
 	logtesting "github.com/knative/pkg/logging/testing"
+
```

Comment:
Remove this newline.

Reply:
working as intended. I have been separating the . imports from the others.

Reply:
What is a ". import"?

Reply:
what you are commenting on: `. "github.com/knative/eventing/pkg/reconciler/testing"`

Reply:
So ". import" is one without an explicit alias? E.g. "github.com/foo/bar" as opposed to "qux github.com/foo/bar"?

Reply:
it imports the exports from that package without a package. 

so: 

`import qux github.com/foo/bar`

`qux.Thing()`

vs

`import . github.com/foo/bar`

`Thing()`

Good for testing, don't do this in prod code.

Reply:
Interesting, thanks for explaining. 

Not part of the PR, but how about removing the newline on line 31?

Reply:
That line breaks the `k8s.io` into a single block.

Reply:
Also you can do this: `import _ "github.com/foo/goo"` and it will not import anything but run the init func and maybe that mutates global state.

Reply:
The project should have some well defined standard ordering. The one I have been using is standard libraries -> new line -> all others. I can see the advantage of moving ". imports" and possibly "_ imports" into separate structures, do we want the k8s stuff separate too? If we do that, then do we want the knative stuff separate as well?

Reply:
Most go projects will separate in the following ways:

- system stuff
- blocks of dep imports that are from the same host *n
- project stuff

Reply:
serving tends to do this, k8s.io gets its own block: https://github.com/knative/serving/blob/master/cmd/controller/main.go#L23

Reply:
I think dot imports should be separated from other imports, since the reader **must** know about them to read the code. I don't have strong opinions on import ordering otherwise. `gofmt` already enforces the block of standard lib, so that's settled. I could understand having separate blocks for vendored code versus in-repo code.

IMO dot imports reduce readability of the code by adding magic. I'd prefer that these not be used, but I won't block the PR for them. Correct me if I'm wrong @n3wscott, but I believe they're used here to work around the fact that when functional options are used, each functional option requires a reference to its package. Without dot imports this would be verbose.


```go
fixtures.NewSubscription(subscriptionName, testNS,
  fixtures.WithSubscriptionChannel(channelGVK, channelName),
  fixtures.WithSubscriptionSubscriberRef(subscriberGVK, subscriberName),
)
```

I prefer builders for specifying fixtures. Builders require only one package reference. When using a builder, my IDE can offer concise suggestions of which builder methods can be called next. With functional options, this is more difficult because any function in scope could conceivably be called (I'm not sure if any IDE is sophisticated enough to filter only the functions with the correct signature).


```go
fixtures.NewSubscription(subscriptionName, testNS).
  WithSubscriptionChannel(channelGVK, channelName).
  WithSubscriptionSubscriberRef(subscriberGVK, subscriberName)
```

Builders have the disadvantage of not returning a concrete struct of their built type, but this can be worked around by embedding the built struct into the builder struct.

Functional options have the advantage of allowing certain options to act on any `runtime.Object` (as pointed out by @n3wscott), which is cool, but seems even harder to document.